velox/functions/sparksql/tests/ArrayMaxTest.cpp

```diff
@@ -0,0 +1,170 @@
+ * limitations under the License.
+ */
+
+#include <type/Timestamp.h>
+#include <limits>
+#include <optional>
+#include "velox/functions/sparksql/tests/SparkFunctionBaseTest.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace facebook::velox::functions::sparksql::test {
+namespace {
+
+class ArrayMaxTest : public SparkFunctionBaseTest {
+ protected:
+  template <typename T>
+  void testExpr(
+      const std::optional<T>& expected,
+      const std::vector<std::optional<T>>& input) {
+    std::vector<std::vector<std::optional<T>>> vec = {input};
+    auto row = makeRowVector({makeNullableArrayVector(vec)});
+    auto result = evaluateOnce<T>("array_max(C0)", row);
+    EXPECT_EQ(expected, result);
+  }
+};
+
+TEST_F(ArrayMaxTest, boolean) {
+  testExpr<bool>(true, {true, false});
+  testExpr<bool>(true, {true});
+  testExpr<bool>(false, {false});
+  testExpr<bool>(std::nullopt, {});
+  testExpr<bool>(true, {true, false, true, std::nullopt});
+  testExpr<bool>(true, {std::nullopt, true, false, true});
+  testExpr<bool>(false, {false, false, false});
+  testExpr<bool>(true, {true, true, true});
+}
+
+TEST_F(ArrayMaxTest, varchar) {
+  testExpr<StringView>("red"_sv, {"red"_sv, "blue"_sv});
+  testExpr<StringView>(
+      "yellow"_sv, {std::nullopt, "blue"_sv, "yellow"_sv, "orange"_sv});
+  testExpr<StringView>(std::nullopt, {});
+  testExpr<StringView>(std::nullopt, {std::nullopt});
+}
+
+// Test non-inlined (> 12 length) nullable strings.
+TEST_F(ArrayMaxTest, longVarchar) {
+  testExpr<StringView>(
+      "red shiny car ahead"_sv,
+      {"red shiny car ahead"_sv, "blue clear sky above"_sv});
+  testExpr<StringView>(
+      "yellow rose flowers"_sv,
+      {std::nullopt,
+       "blue clear sky above"_sv,
+       "yellow rose flowers"_sv,
+       "orange beautiful sunset"_sv});
+  testExpr<StringView>(std::nullopt, {});
+  testExpr<StringView>(
+      "red shiny car ahead"_sv,
+      {"red shiny car ahead"_sv,
+       "purple is an elegant color"_sv,
+       "green plants make us happy"_sv});
+}
+
+TEST_F(ArrayMaxTest, date) {
+  auto D = [](const std::string& dateStr) { return DATE()->toDays(dateStr); };
+  testExpr<int32_t>(D("2023-08-23"), {D("1970-01-01"), D("2023-08-23")});
+  testExpr<int32_t>(std::nullopt, {});
+  testExpr<int32_t>(D("1970-01-01"), {D("1970-01-01"), std::nullopt});
+}
+
+TEST_F(ArrayMaxTest, timestamp) {
+  auto T = [](int64_t micros) { return Timestamp::fromMicros(micros); };
+  testExpr<Timestamp>(T(1), {T(0), T(1)});
+  testExpr<Timestamp>(
+      Timestamp::max(), {T(0), T(1), Timestamp::max(), Timestamp::min()});
+  testExpr<Timestamp>(std::nullopt, {});
+  testExpr<Timestamp>(T(0), {T(0), std::nullopt});
+}
+
+template <typename Type>
+class ArrayMaxIntegralTest : public ArrayMaxTest {
+ public:
+  using T = typename Type::NativeType::NativeType;
+
+  void runTest() {
```

Comment:
This method is used only once. Consider folding it into the caller.

Reply:
sorry, `runTest` uses `T`, which is not accessible in the caller.

Reply:
> sorry, runTest uses T, which is not accessible in the caller.

Is this so? I believe T should be accessible to the caller.

Reply:
Hi @mbasmanova, I just moved `runTest` code outside of the class and place it into test caller. Thanks!